### PR TITLE
[MLIR] Handle exceptions

### PR DESCRIPTION
**Pull request by erick-xanadu**
**Context:**  Exceptions raised during run time would trigger segmantation faults.

**Description of the Change:** Adds MLIR pass to check for errors returned from asynchronous functions

**Benefits:** No segfaults.

**Possible Drawbacks:** One thing I noticed is that even without the exception handling transformation, it looks like when asynchronous execution happens in parallel pytests, there is a non-zero probability of failure. This gets reduced the less number of parallel tests are run. I will ask the community and see if there is a way to further diagnose this issue.

**Related GitHub Issues:** This PR does not implement the transformation that would propagate the exception handling algorithm up the call graph. This is because inlining guarantees that all calls to quantum operations happen in the "main" method. But this is something that we might consider in the future, or do the runtime approach instead. Things that are left as TODOs:

* Use a stack and a mutex to preserve error messages from multiple threads
*  Before this PR, the runtime did not raised errors when `qml.CNOT(wires=[x, y])` given `x == y`. There might be other places where we need to raise these exceptions.
 * De-allocate device and quantum register when an exception happens.
 
[sc-52638]

### CODE REVIEW

**Comment by rauletorresc** [comment]
I see some opportunities to improve this PR:

1. Squash all documentation commits into one, as having them separate does not bring any benefit, instead it would make the commit history longer.
2. I think the style changes should either be squashed into the previous commit where the style was broken, or all the style changes should be squashed into one.
3. Most commits have a one or two words description, this will make it difficult to track the intent of the commit in the future. Isn't this a sign that some commits can be merge into one, which can have a meaningful description?
4. Shouldn't we provide some kind of label to differentiate commits making changes to the code logic from commits that don't do it, like documentation or style commits?
5. I noticed some commits with the description "documentation" have actually made small changes to the code logic. I think those two actions should be split into two different commits.

File: mlir/lib/Catalyst/Transforms/DetectQNodes.cpp
Comment (rauletorresc) [inline mlir/lib/Catalyst/Transforms/DetectQNodes.cpp:203]: The "match" method changed its implementation. I would suggest moving this change to a different commit, because the commit description says it is related to "Documentation".

File: mlir/lib/Catalyst/Transforms/DetectQNodes.cpp
Comment (rauletorresc) [inline mlir/lib/Catalyst/Transforms/DetectQNodes.cpp]: The deletion of this "return" should be placed in another commit, as it is not related to "Documentation".

File: mlir/lib/Catalyst/Transforms/DetectQNodes.cpp
Comment (rauletorresc) [inline mlir/lib/Catalyst/Transforms/DetectQNodes.cpp:329]: This additional instruction should be placed in a separate commit, as it is not related to "Documentation"

File: mlir/lib/Catalyst/Transforms/DetectQNodes.cpp
Comment (rauletorresc) [inline mlir/lib/Catalyst/Transforms/DetectQNodes.cpp:116]: This set of functions were not declared here before. This does not seem to be a "Documentation" change. I would suggest making a different commit for those kind of changes.

File: mlir/lib/Catalyst/Transforms/DetectQNodes.cpp
Comment (rauletorresc) [inline mlir/lib/Catalyst/Transforms/DetectQNodes.cpp:844]: This change affects the logic of the program and must be placed in a different commit that is not documentation

File: mlir/test/Catalyst/AsyncException.mlir
Comment (rauletorresc) [inline mlir/test/Catalyst/AsyncException.mlir:219]: This change affects the logic of the program and must be placed in a different commit that is not documentation

**Comment by josh146** [comment]
@rauletorresc just for some context, our current strategy for merging PRs is a merge commit, with the commit title/description automatically replaced with the PR title and description :) So commits from a merged branch are not preserved in `main`. 

This also encourages us to have useful and detailed PR descriptions.

**Review by rauletorresc — CHANGES_REQUESTED** [review]
See my comments here: https://github.com/PennyLaneAI/catalyst/pull/447#issuecomment-1892405503

File: mlir/lib/Catalyst/Transforms/DetectQNodes.cpp
Comment (erick-xanadu) [reply mlir/lib/Catalyst/Transforms/DetectQNodes.cpp:203]: Yes, due to the previous pass, which annotates:

```
        candidate &&
        hasQnodeAttribute(candidate.value()) 
 ```
 
 with `isScheduledForTransformation` the previous two checks are unnecessary.
 
 I can split the commit if you want.

**Comment by rauletorresc** [comment]
> @rauletorresc just for some context, our current strategy for merging PRs is a merge commit, with the commit title/description automatically replaced with the PR title and description :) So commits from a merged branch are not preserved in `main`.
> 
> This also encourages us to have useful and detailed PR descriptions.

Thanks for the clarification! Still, a cleaner commit history would ease the code review process at least. 

In the case of a merged PR, if this is too big, it might make it a bit difficult to track a bug introduced in one of the merged small commits. Just a thought.

File: mlir/lib/Catalyst/Transforms/DetectQNodes.cpp
Comment (erick-xanadu) [reply mlir/lib/Catalyst/Transforms/DetectQNodes.cpp:116]: I am moving functions such that the documentation can read from top to bottom. In order to have the code compile, I just needed to add declaration for functions.

File: mlir/lib/Catalyst/Transforms/DetectQNodes.cpp
Comment (erick-xanadu) [reply mlir/lib/Catalyst/Transforms/DetectQNodes.cpp:329]: I moved this function call from before the if statement to after the if statement. I just wanted the annotation for the next pass to be the last change in this function. This is also why I removed the return statement from the if statement.

File: mlir/lib/Catalyst/Transforms/DetectQNodes.cpp
Comment (rauletorresc) [reply mlir/lib/Catalyst/Transforms/DetectQNodes.cpp:203]: No need now. I just think separating the changes according to their nature would ease our review process.

File: mlir/lib/Catalyst/Transforms/DetectQNodes.cpp
Comment (rauletorresc) [reply mlir/lib/Catalyst/Transforms/DetectQNodes.cpp:329]: Thanks for clarifying. This clarification could have be a good commit description if it was made in a separate commit. As I am new at the code base, it is difficult for me to understand the motivation even for a minor change.

**Review by rauletorresc — APPROVED** [review]

File: runtime/lib/capi/RuntimeCAPI.cpp
Comment (maliasadi) [inline runtime/lib/capi/RuntimeCAPI.cpp:438]: Thanks! Would you be able to add a unit test for this check in `Test_LightningCoreQIS.cpp` for the coverage report?

File: runtime/lib/capi/RuntimeCAPI.cpp
Comment (maliasadi) [inline runtime/lib/capi/RuntimeCAPI.cpp:84]: Would you add a simple test for this new method in `Test_LightningCoreQIS.cpp` for the codecov report?

File: frontend/test/async_tests/test_async.py
Comment (maliasadi) [inline frontend/test/async_tests/test_async.py:344]: :raised_hands:

File: frontend/test/async_tests/test_async.py
Comment (maliasadi) [inline frontend/test/async_tests/test_async.py:323]: would you think it's necessary to add a test for `catalyst.for_loop`?

File: frontend/test/async_tests/test_async.py
Comment (maliasadi) [inline frontend/test/async_tests/test_async.py:370]: can we add checks within async execution of multiple qnodes in calls to `catalyst.adjoint` and `catalyst.ctrl`?

File: frontend/test/async_tests/test_async.py
Comment (maliasadi) [inline frontend/test/async_tests/test_async.py:371]: Do you aware of any corner cases for this exception handling support?

File: mlir/lib/Catalyst/Transforms/DetectQNodes.cpp
Comment (maliasadi) [inline mlir/lib/Catalyst/Transforms/DetectQNodes.cpp:32]: nice! didn't know that you can use `llvm::StringRef` with `constexpr` when building with C++17. It's probably implemented as `std::string_view` instead of `std::string` :thinking: as you can use `constexpr` with `string_view` in C++17. `constexpr std::string` is a C++20 feature.

File: mlir/lib/Catalyst/Transforms/DetectQNodes.cpp
Comment (maliasadi) [inline mlir/lib/Catalyst/Transforms/DetectQNodes.cpp:1212]: consider this suggestion for the code consistency:
 ```suggestion
    if (!funcOp->hasAttr(passthroughAttr)) {
        return false;
    }
```

File: mlir/lib/Catalyst/Transforms/DetectQNodes.cpp
Comment (maliasadi) [inline mlir/lib/Catalyst/Transforms/DetectQNodes.cpp:1219]:
```suggestion
    auto it = std::find(haystack.begin(), haystack.end(), needle);
    return (it != haystack.end()) ? true : false;
```

File: mlir/lib/Catalyst/Transforms/DetectQNodes.cpp
Comment (maliasadi) [inline mlir/lib/Catalyst/Transforms/DetectQNodes.cpp:1221]:
```suggestion
```

File: mlir/lib/Catalyst/Transforms/DetectQNodes.cpp
Comment (maliasadi) [inline mlir/lib/Catalyst/Transforms/DetectQNodes.cpp:91]: Would you like to consider inlining these helper methods? Would you be able to add them in a header file?

File: mlir/lib/Catalyst/Transforms/DetectQNodes.cpp
Comment (maliasadi) [inline mlir/lib/Catalyst/Transforms/DetectQNodes.cpp:131]: please consider moving these helper functions to a different file.

File: runtime/lib/capi/RuntimeCAPI.cpp
Comment (maliasadi) [inline runtime/lib/capi/RuntimeCAPI.cpp:84]: If this function is currently called only within the context of async qnodes, will you interested in updating the message to better convey the association with the issue? 

Maybe :thinking: 
```suggestion
void __catalyst__host__rt__unrecoverable_error() { RT_FAIL("Unrecoverable error from asynchronous execution of multiple quantum programs."); }
```

**Review by maliasadi — CHANGES_REQUESTED** [review]
Great work with sufficient amount of code comments :raccoon: I've had a few suggestions to address and then we can merge it :rocket:

**Comment by maliasadi** [comment]
Don't forget to update the changelog :) You can also add GH label: `ci:build_wheels` after addressing those comments to check wheels before merge.

File: mlir/test/Catalyst/AsyncException.mlir
Comment (erick-xanadu) [reply mlir/test/Catalyst/AsyncException.mlir:221]: Thanks! I'll try to improve them too. There is a lot of room to make them better. :sweat_smile:

Comment (erick-xanadu) [reply mlir/test/Catalyst/AsyncException.mlir]: I've improved the tests quite a bit. I am resolving this issue. Feel free to reopen :)

File: mlir/include/Catalyst/Transforms/Passes.td
Comment (erick-xanadu) [reply mlir/include/Catalyst/Transforms/Passes.td]: https://github.com/PennyLaneAI/catalyst/pull/447/commits/44ef755c764d36442494d403125bf210aa8ace0b

File: frontend/test/async_tests/test_async.py
Comment (erick-xanadu) [reply frontend/test/async_tests/test_async.py:1]: https://github.com/PennyLaneAI/catalyst/issues/459 I'll resolve this comment, but feel free to reopen.

File: runtime/lib/capi/RuntimeCAPI.cpp
Comment (erick-xanadu) [reply runtime/lib/capi/RuntimeCAPI.cpp]: https://github.com/PennyLaneAI/catalyst/pull/447/commits/372f237f8f9c38e548794aaf40bce1d286c86817

File: runtime/lib/capi/RuntimeCAPI.cpp
Comment (erick-xanadu) [reply runtime/lib/capi/RuntimeCAPI.cpp:441]: https://github.com/PennyLaneAI/catalyst/pull/447/commits/f156f31b67feb2db2090b4c3b8d8e3a0cb329e8e

File: frontend/test/async_tests/test_async.py
Comment (erick-xanadu) [reply frontend/test/async_tests/test_async.py:371]: Yes, if we have a more complex call graph, we haven't handle that. E.g.,

If we have the following callgraph

```
                      A
                  /         \
                 B         C
```

And exceptions happened in either B or C, then we would be able to handle it. However, if exceptions happened at different levels in the callgraph. E.g.,

```
                      A
                  /         \
                 B         C
                 |
                 D
```

If an exception happened in D, and C is executing asynchronously, then we would exit without waiting for the values of C. The solution is to propagate this analysis up the stack. And have the `llvm.call` to A to B be a `llvm.invoke`. The other alternative would be to do something more on the runtime.

Since the inliner inlines everything except functions containing quantum operations, I believe we are guaranteed to have only callgraphs of the above type.

But this is only the first step, so we can always work to expand support.

File: mlir/lib/Catalyst/Transforms/DetectQNodes.cpp
Comment (erick-xanadu) [reply mlir/lib/Catalyst/Transforms/DetectQNodes.cpp:91]: I prefer to let the compiler decide what to inline. :)

File: mlir/lib/Catalyst/Transforms/DetectQNodes.cpp
Comment (erick-xanadu) [reply mlir/lib/Catalyst/Transforms/DetectQNodes.cpp:57]: Moved them in several PRs.

File: frontend/test/async_tests/test_async.py
Comment (erick-xanadu) [reply frontend/test/async_tests/test_async.py:370]: I'll add it, but it shouldn't matter which operation is in the program. The exception handling code happens in the caller of the qnode.

Comment (erick-xanadu) [reply frontend/test/async_tests/test_async.py:416]: Here's adjoint: https://github.com/PennyLaneAI/catalyst/pull/447/commits/402a4555c3f6cb30e7cbcdf62b711468a1048d43

File: frontend/test/async_tests/test_async.py
Comment (erick-xanadu) [reply frontend/test/async_tests/test_async.py:351]: https://github.com/PennyLaneAI/catalyst/pull/447/commits/a2fc25d46deb8d40b417497157a96d69feccb01f

File: runtime/tests/Test_LightningCoreQIS.cpp
Comment (erick-xanadu) [inline runtime/tests/Test_LightningCoreQIS.cpp:1902]: @maliasadi it is still here. You might be looking at an old version?

File: frontend/test/async_tests/test_async.py
Comment (erick-xanadu) [reply frontend/test/async_tests/test_async.py:105]: It is planned for the second PR. The idea is to propagate the error message from all async circuits (or can we can discuss and settle on something else).

**Review by maliasadi — APPROVED** [review]
Nice work! :1st_place_medal:

The Python Formatter action is still failing, but everything else looks good.

**Comment by erick-xanadu** [comment]
> The Python Formatter action is still failing, but everything else looks good.

Can't do anything about that. It only complains because I moved files and moved them back. We keep this one because otherwise we have the bug:

```
--- /home/runner/work/catalyst/catalyst/frontend/test/conftest.py:before	2024-02-13 01:09:28.106767
+++ /home/runner/work/catalyst/catalyst/frontend/test/conftest.py:after	2024-02-13 01:09:30.576114
@@ -18,8 +18,9 @@
 import pytest
 
 try:
+    import tensorflow as tf
+
     import catalyst
-    import tensorflow as tf
 except (ImportError, ModuleNotFoundError) as e:
     tf_available = False
 else:
```

And codefactor complains about files I never touched.

**Comment by erick-xanadu** [comment]
@dime10 or @maliasadi could you merge this? I can't merge it due to the code factor and python formatting.

File: frontend/test/async_tests/__init__.py
Comment (dime10) [inline frontend/test/async_tests/__init__.py:1]: I think we can remove this file no?

Comment (erick-xanadu) [reply frontend/test/async_tests/__init__.py:1]: Yes, you are right. Thanks! https://github.com/PennyLaneAI/catalyst/pull/447/commits/c75e207b77ad643dfb1099eb026d12b898f99890

**Pull request merged**
